File: javascript/binary-conversions.js
function BinaryConverter(binNum) {
  var numArray = binNum.toString().split("").reverse();
  var number = 0;
  for (var i=0; i < numArray.length; i++) {
    var num = parseInt(numArray[i]);
    var location = Math.pow(2, i);
    number += num * location;
  }
  return number;
}

console.log(BinaryConverter(101)) // 5
console.log(BinaryConverter(1000)) // 8

///////////////////////////////////////////////////////////////////

function toBinaryNumber(number) {
  var returning = new Array;
  if (number % 2 != 0) {
    returning[0] = 1;
    number -= 1;
  }

  while (number > 0) {
    i = 1;
    while (Math.pow(2, i) <= number) {
      if (Math.pow(2, (i + 1)) > number) {
        returning[i] = 1;
        number -= (Math.pow(2, i));
        break;
      }
      i++;
    }
  }

  for (var j=0; j < returning.length; j++) {
    if (returning[j] == null) {
      returning[j] = "0";
    } else {
      returning[j] = returning[j].toString();
    }
  }

  return parseInt(returning.reverse().join(""));
}

console.log(toBinaryNumber(8) === 1000);

File: Ruby/binary-conversions.rb
def from_binary(number)
  number_array = number.to_s.chars.reverse
  returning = 0
  number_array.each_with_index do |item, i|
    location = 2 ** i
    returning += (location * item.to_i)
  end
  returning
end

# puts from_binary(101) # 5
# puts from_binary(1000) # 8
puts from_binary(11001)

def to_binary(number)
  returning = []

  if number.odd?
    returning << 1
    number -= 1
  end

  while number > 0
    i = 1
    while 2 ** i <= number
      if 2 ** (i + 1) > number
        returning[i] = 1
        number -= (2 ** i)
        break
      end
      i += 1
    end
  end
  returning.map{|num| num.nil? ? "0" : "#{num}"}.reverse.join.to_i
end

# puts to_binary(9)
